src/types.ts

```diff
@@ -1063,7 +1063,7 @@ export class Signature {
     } else {
       // check `this` type (invariant)
       if (thisThisType) {
-        if (targetThisType != targetThisType) return false;
+        if (targetThisType == null) return false;
```

Comment:
Shouldn't this be:
```suggestion
        if (targetThisType == thisThisType) return false;
```

Reply:
No, target type can be child class of this type. In this case,  `if (targetThisType != thisThisType)` lead to wrong `false`

Reply:
Looks like `targetThisType != thisThisType` is what's intended here, though. Perhaps what this is missing is a special case for overloads?

Reply:
I have no idea about overloads, but I do recall TS being strict in some way about function types being assignable. Something like `(this: B) => void` can't be casted to `(this: A) => void` if `B extends A`, since a different subtype of `A` could be passed to the target function when called.

Instead, I think `(this: A) => void` can be casted to `(this: B) => void` IIRC? So `targetThisType.isAssignableTo(thisThisType)` might be the right answer?

Reply:
Typescript in different with our current case. In AssemblyScript, `this` is a implicit first parameter of the function. In typescript/javascript, `this` is a search chain which can not only refer to the object itself, but may also refer to global/window object. The bug this PR want to fix is due to the `this` type checking, so that we don't need to refer to typescript design.

It seems not possible to create a testcase that cast  `(this: B) => void` to `(this: A) => void`. For example, in following code, the thisThisType is `a\A`, the targetThisType is null.
```typescript
class A{
  public foo():void{}
}

class B extends A{
}

function expectCallback(callback:(arg1: A)=>void):void{

}

export function fut(): void {
  const b = new B();
  expectCallback(b.foo);
}

fut();

```

Reply:
I just tried that, and it seems that the compiler errors on method declarations that override those from implemented interfaces...is that what you meant, @dcodeIO?

Reply:
> so that we don't need to refer to typescript design.

Whoops, sorry, I didn't see your comment. We should be mimicking TS semantics (and it shouldn't hurt to).

> but may also refer to global/window object.

This isn't really a concern, since AssemblyScript code is ESM, and in regular JS/TS, ESM implies strict mode, which means `this` cannot refer to the global/window object.

> It seems not possible to create a testcase that cast `(this: B) => void` to `(this: A) => void`.

That *should* throw an error. But `(this: A) => void` to `(this: B) => void` *should* work just fine. [This](https://www.assemblyscript.org/editor#IyFvcHRpbWl6ZT1zcGVlZCZydW50aW1lPWluY3JlbWVudGFsCmNsYXNzIEEgewogICAgZm9vKCk6IHZvaWQgewogICAgICAgIGNvbnNvbGUubG9nKCJBIikKICAgIH0KfQoKY2xhc3MgQiBleHRlbmRzIEEgewogICAgZm9vKCk6IHZvaWQgewogICAgICAgIGNvbnNvbGUubG9nKCJCIikKICAgIH0KfQoKZXhwb3J0IGZ1bmN0aW9uIHRlc3QoKTogdm9pZCB7IAogIGNvbnN0IGEgPSBuZXcgQSgpCiAgY29uc3QgYiA9IG5ldyBCKCkKCiAgYS5mb28uY2FsbChiKSAvLyBzaG91bGQgbG9nICJBIgp9CiMhaHRtbAo8c2NyaXB0IHR5cGU9Im1vZHVsZSI+CiAgICBjb25zdCB7dGVzdH0gPSBhd2FpdCBpbnN0YW50aWF0ZShhd2FpdCBjb21waWxlKCkpCiAgICB0ZXN0KCkKPC9zY3JpcHQ+) should do it.

By playing around with that, it seems like `b.foo.call(a)` errors, but `(b.foo as (this: A) => void).call(a)` compiles when it shouldn't. On the other hand, `a.foo.call(b)` and `(a.foo as (this: B) => void).call(b)` both compile as expected. And in the end, it doesn't even matter, because the compiled output seems to log "B" no matter what...

I'm not sure what to think.

Reply:
`(b.foo as (this: A) => void).call(a)`. is out of scope of this ticket. I just tried to comiple follow example, and the `Signature.isAssignableTo` function is not called. So that the problem you just mentioned should not relavet to current ticket. If we want to fix that, we would better create another ticket to track it.
```typescript
class A {
  public foo(): void {}
}

class B extends A {}

export function fut(): void {
  const a = new A();
  const b = new B();
  (b.foo as (this: A) => void).call(a);
}
```

About the current ticket, it can be summraized as persudo code:
```
if(thisThisType == null && targetThisType == null){
check reset parameters
}else if((thisThisType == null && targetThisType != null) || (thisThisType != null && targetThisType == null)){
return false;
}else{
Not such test case yet.
}
```
Currently there still no test case under `tests` folder, where `thisThisType != null && targetThisType != null`. 
The consern you mentioned are also not that case.

Reply:
@dcodeIO @CountBleck 
I just updated the code, there a 3 cases. 
1. `targetThisType == null && thisThisTyps == null`. In this case, just check further other parameters.
2. `(targetThisType != null && thisThisTyps == null) || (targetThisType == null && thisThisTyps != null)`. In this case, just return false.
3. `targetThisType != null && thisThisTyps != null`. This case only happens when `finalizeOverrideStub`. So that `thisThisType.isAssignableTo(targetThisType)` shall be always true. I just added an assert there.

Moreover, the `(b.foo as (this: A) => void).call(a)` is another case out of this ticket's scope. I tired to compile following sample, but it never used `Signature.isAssignableTo`. If this needs to be fixed, we should create another ticket to track it.
```typescript
class A {
  public foo(): void {}
}

class B extends A {}

export function fut(): void {
  const a = new A();
  const b = new B();
  (b.foo as (this: A) => void).call(a);
}
```

Reply:
The existing code already handled case 1 and case 2 (just in a more concise manner).

> I tired to compile following sample, but it never used `Signature.isAssignableTo`.

I didn't realize that, whoops! :D
The compiler does crash from your assert by the way:
```ts
class A {
  foo(): void {
    console.log("A");
  }
}

class B extends A {
  foo(): void {
    console.log("B");
  }
}

function foo(): void {
  console.log("nothing");
}

function consume(callback: (this: B) => void): void {
  const b = new B();
  callback.call(b);
}

export function testNull(): void { 
  consume(foo); // This should (and does) error; this is fine.
}

export function testA(): void {
  const a = new A();
  consume(a.foo); // This shouldn't error...but instead it crashes!
}
```

I don't have time at the moment, but I'll reply later with a better example for what I was talking about earlier...

Reply:
@CountBleck 
Yes, case 1 and case 2 are clear and easy to handle. The only problem is case 3. In a previous comment, you mentioned to check if `thisThisType` is assignable to `targetThisType`, I tried as your comment. The conclusion is, base on current test cases, the thisThisType is always child class of targetThisType, so that it doesn't need to be checked. Then I just changed the `if(thisThisType.isAssignableTo(targetThisType))` to assert. 
The assert shall not crash, otherwise it can't pass CI checking.

Reply:
The snippet above crashes the compiler due to your assert, and that should not happen.

Reply:
@CountBleck Thank you for your help. I sorry I misunderstood your example a bit yesterday, but now I reproduced this crash and fixed it in a new commit. And I have also updated your example into test case.
both `thisThisType.isAssignableTo(targetThisType)` and `targetThisType.isAssignableTo(thisThisType)` need to be checked. The first one for override, the second one for type casting as your example.

Does it looks fine now?

Reply:
Thanks to @HerrCai0907, the `thisThisType.isAssignableTo(targetThisType)` should not be checked in `Signature.isAssignableTo`. Instead, that's caused a bug of `finalizeOverrideStub` in `compiler.ts`.

Reply:
> Does it looks fine now?

Not quite. I'll do another review...

Reply:
@CountBleck Your comments are fixed in new commit

Reply:
@CountBleck format and style are refined